R/utilities-simulation-results.R

```diff
@@ -0,0 +1,126 @@
+#' @param individualIds \code{numeric} IDs of individiuals for which the results should be extracted.
+#' By default, all individuals from the results are considered. If the individual with the provided ID is not found, the ID is ignored
+#'
+#' @return A list of results, each entry having the name of a quantity path and consisting of the element \code{x} being the time values (in minutes)
+#' and the element \code{y} a \code{NxM} matrix of simulated values, with \code{N} the number of entries (equals the number of output time steps)
+#' and \code{M} the number of individuals. If no individual contains the results for a given path, the value of the element is \code{NULL}
+#'
+#' @seealso \code{\link{getResultsValues}}
+#'
+#' @export
+#'
+#' @examples
+#' simPath <- system.file("extdata", "simple.pkml", package = "ospsuite")
+#' sim <- loadSimulation(simPath)
+#' simResults <- runSimulation(sim)
+#'
+#' paths <- "Organism|Volume"
+#'
+#' resultsValues <- getPopulationResultsValues(simResults, paths, individualIds = 0)
+#' plot(resultsValues[[paths]])
+getPopulationResultsValues <- function(simulationResults, quantitiesOrPaths, individualIds = NULL) {
+  validateIsOfType(simulationResults, "SimulationResults")
+  quantitiesOrPaths <- c(quantitiesOrPaths)
+  validateIsOfType(quantitiesOrPaths, c("Quantity", "character"))
+  validateIsNumeric(individualIds, nullAllowed = TRUE)
+
+  # If quantities are passed, get their paths.
+  paths <- quantitiesOrPaths
+  if (isOfType(quantitiesOrPaths, "Quantity")) {
+    paths <- unlist(lapply(quantitiesOrPaths, function(x) x$consolidatePath))
+  }
+  paths <- unique(paths)
+
+  # If no specific individual ids are passed, iterate through all individuals
+  if (is.null(individualIds)) {
+    individualIds <- simulationResults$allIndividualIds
+  }
+  else {
+    individualIds <- unique(individualIds)
+  }
+
+  # Pre-alloate a list with the number of entries equal to the number of paths
+  output <- vector("list", length(paths))
+  names(output) <- paths
+  # All time values are equal
+  timeVals <- simulationResults$timeValues
+
+  for (path in paths) {
+    # Pre-allocate the output matrix. The number of rows equals the number of simulated time points.
+    # The number of columns is the number of possible individuals.
+    outputValues <- matrix(nrow = length(timeVals), ncol = max(length(individualIds), simulationResults$count), dimnames = NULL)
+
+    for (individualIdx in seq_along(individualIds)) {
+      individualId <- individualIds[individualIdx]
+      vals <- simulationResults$getValuesForIndividual(resultPath = path, individualId = individualId)
+      if (is.null(vals)) {
+        next
+      }
+      outputValues[, individualIdx] <- vals
+    }
+    # Remove all NAs, i.e. columns for individuals that are not present in the
+    # results. Solution from https://github.com/sfirke/janitor
+    mask_keep <- colSums(!is.na(outputValues)) > 0

```

Comment:
I first allocate the matrix with the number of colums equal to the number of passed individual ids,  but limited by the total number of simulated individuals. So allocating with the most possible size. I did this because, if I understand it correctly, it prevents the matrix being copied every time it is changed in the loop, which would make the loop potentially very slow.

After filling the matrix, it is reduced to the dimensions equal to the number of individuals the results were found for. So yes, this will most probably happen only if the user passes wrong ids.